frappe/core/doctype/user/user.py

```diff
@@ -728,8 +730,14 @@ def _get_user_for_update_password(key, old_password):
 	# verify old password
 	result = frappe._dict()
 	if key:
-		result.user = frappe.db.get_value("User", {"reset_password_key": key})
-		if not result.user:
+		user = frappe.db.get_value("User", {"reset_password_key": key}, ["name", "reset_password_key_datetime"])
+		result.user, reset_password_key_datetime = user if user else (None, None)
+
+		if result.user:
+			reset_password_link_expiry = frappe.db.get_single_value("System Settings", "reset_password_link_expiry_seconds")
+			if reset_password_link_expiry and now_datetime() > reset_password_key_datetime + timedelta(seconds=reset_password_link_expiry):
+				result.message = _("The Link specified has been expired")
```

Comment:
```suggestion
			if reset_password_link_expiry and now_datetime() > reset_password_key_datetime + timedelta(seconds=reset_password_link_expiry):
				frappe.local.response.http_status_code = 410
				result.message = _("The reset password link has been expired.")
```

let's also set the appropriate HTTP status code.

Reply:
The status_code is already set `update_password` which invokes `_get_user_for_update_password`  before returning the final response